### PR TITLE
DBZ-78 add Debezium auto-creation topics configuration documentation

### PR DESCRIPTION
**Pull request by rk3rn3r**
Debezium Configuration docs for https://issues.redhat.com/browse/DBZ-78

### CODE REVIEW

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+topic.creation.enable = true
+----
+
+// TODO: how to express this for downstream?
+[NOTE]
+====
+If you don't want to allow automatic topic creation by connectors you can set this value to `false`
+in the Kafka Connect config (_connect-distributed.properties_ file or via environment variable
+_CONNECT_TOPIC_CREATION_ENABLE_ when using https://hub.docker.com/r/debezium/connect[Debezium's container image for Kafka Connect]).
+====
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: Can we use that for the downstream docs as well or do we need a different wording here?
What do you suggest? @gunnarmorling  @TovaCohen

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+:linkattrs:
+:icons: font
+:source-highlighter: highlight.js
+
+toc::[]
+
+{prodname} automatically creates *internal* topics for offsets, connector status, config
+storage and history topics. The destination topics for the captured tables will be
+automatically created with a default config by the Kafka brokers when
+`auto.create.topics.enable` is enabled.{empty} +
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:19]:
```suggestion
`auto.create.topics.enable` is set to `true`.{empty} +
```

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    "topic.creation.applicationlogs.retention.ms": 7776000000,
+    "topic.creation.applicationlogs.compression.type": "lz4"
+}
+----
+
+ifdef::community[]
+== Additional resources
+- Debezium Blog: https://debezium.io/blog/2020/09/15/debezium-auto-create-topics/[Auto-creating Debezium Change Data Topics]
+- Kafka Improvement Proposal about adding topic auto-creation to Kafka Connect: https://cwiki.apache.org/confluence/display/KAFKA/KIP-158%3A+Kafka+Connect+should+allow+source+connectors+to+set+topic-specific+settings+for+new+topics[KIP-158]
+endif::community[]
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:216]: I only allowed these resources in community. Is that fine or can this be in the downstream docs as well?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 1 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: No, for downstream this should refer to the connector CRD equivalent of this configuration.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+== Configuration
+
+Topic auto-creation is based on groups. Every custom group has an `include` and an
+`exclude` property which are comma-separated lists of regular expressions matching
+topic names that should be included or excluded.
+
+[NOTE]
+====
+You can specify both, `include` and `exclude` parameters, but note that exclusion rules
+have precedent and override any inclusion rules for topics.
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:52]:
```suggestion
take precedence and override any inclusion rules for topics.
```

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+====
+
+You don't have to specify any custom group. When there's no custom group registered or
+the registered group's `include` patterns don't match the topic which is to be created
+then the default config will be used.
+
+You can specify all https://kafka.apache.org/documentation/#topicconfigs[*topic level configuration parameters*]
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:59]: There's a var in antora.yml or similar for the Kafka docs base URL.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+|1
+|`topic.creation.default.replication.factor` defines the replication factor for topics created by
+the default group.
+
+|2
+|`topic.creation.default.partitions` defines the number of partitions for topics created by
+the default group.
+
+|3
+|`topic.creation.default.cleanup.policy` is mapped to the https://kafka.apache.org/documentation/#cleanup.policy[`cleanup.policy`]
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:94]: See above.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+[source,options="nowrap",json]
+----
+{
+    ...
+
+    "topic.creation.default.replication.factor": 3,
+    "topic.creation.default.partitions": 10,
+    "topic.creation.default.cleanup.policy": "compact",
+    "topic.creation.default.compression.type": "lz4"
+    "topic.creation.groups": "inventory,applicationlogs",
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:196]: Maybe add some empty lines between the groups?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    "topic.creation.applicationlogs.replication.factor": 1,
+    "topic.creation.applicationlogs.partitions": 20,
+    "topic.creation.applicationlogs.cleanup.policy": "delete",
+    "topic.creation.applicationlogs.retention.ms": 7776000000,
+    "topic.creation.applicationlogs.compression.type": "lz4"
+}
+----
+
+ifdef::community[]
+== Additional resources
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:213]: An introductory sentence would be good: "To learn more about topic creation settings...".

**Review by gunnarmorling — COMMENTED** [review]
Nice work, a few comments inline. Can you also add a link to this page somewhere on https://debezium.io/documentation/reference/install.html#_configuring_debezium_topics.

@TovaCohen, your feedback would be welcomed very much. No immediate work for downstream is needed IMO, as there we don't use AK 2.6 yet. But can you please log a FUSEDOC issue already and preliminarily assign it to the Q4 release bucket, so we won't forget about this. Thx!

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 7 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:196]: There are empty lines in the above separate examples. In this complete example I would like to have it very authentic, means all together.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 8 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:213]: very good point! will do that.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 1 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: Do you have an example where we link to a CRD (parallel to the docker hub link)

Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: is it `{DockerKafkaConnect}` ?

Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: See any connector doc page (under "example configuration").

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 3 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:216]: No, we typically don't link to blogs/upstream from prod docs.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 7 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:196]: You can have empty lines in your real-world JSON/CRD, too ;)

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 3 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:223]: So yes, it's fine then?

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 2 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc]: Done

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 4 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc]: Done

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 5 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc]: Done

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 7 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:199]: I know. But don't! :D

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 8 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:216]: Done

File: documentation/modules/ROOT/pages/install.adoc [hunk 9]
```diff
@@ -103,6 +103,10 @@ e.g. when updating them
 ** Replicated in production
 ** Single partition
 *** You can relax the single partition rule but your application must handle out-of-order events for different rows in database (events for a single row are still totally ordered). If multiple partitions are used, Kafka will determine the partition by hashing the key by default. Other partition strategies require using SMTs to set the partition number for each record.
+// the condition can be removed once downstream  is updated to Kafka 2.6+
+ifdef::community[]
+** For customizable topic auto-creation (available since Kafka Connect 2.6.0) see {link-prefix}:{link-topic-auto-creation}[Custom Topic Auto-Creation]
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/install.adoc:108]: added a reference / backlink and note here as well @gunnarmorling

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 1 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:41]: @gunnarmorling This is a CRD for a connector. The config needs to go to the Kafka Connect container env var.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+metadata:
+  name: my-connect-cluster
+
+...
+
+spec:
+  config:
+    topic.creation.enable: "false"
+----
+endif::product[]
```
Comment (rk3rn3r) [inline documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:64]: is the above a correct example for it? WDYT? @gunnarmorling

**Comment by rk3rn3r** [comment]
I just opened this PR for review and merge, but I need to build the docs and test them now prior merging.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 1 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:49]: see below

**Comment by rk3rn3r** [comment]
Looks good  in community view. Do I have to test downstream documentation? If so, how?

Do we need CRD connector config for the last paragraph for downstream or is that JSON sufficient?

**Comment by TovaCohen** [comment]
Here's the FUSEDOC Jira that I created to update this upstream doc, and add what's needed downstream:
https://issues.redhat.com/browse/FUSEDOC-4157
The fix version for this FUSEDOC Jira is Integration-2020-Q4.
I'm afraid I don't understand enough to suggest any updates now, but since we don't yet need updates for downstream, this should not hold up this PR.

**Comment by rk3rn3r** [comment]
@gunnarmorling I added the fallback behavior and default value description based on last night's twitter. ;)

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 10 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:64]: In general, yes, it is. Problem though is that downstream isn't on AK 2.6 yet, so it doesn't really make sense to have this here yet. But it's not a problem, as this page will not be added to downstream docs without explicit opt-in. Can you please log a Jira issue for tracking that for 1.3. Thanks!

File: documentation/modules/ROOT/pages/install.adoc (hunk 9 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/install.adoc:108]: +1

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 3 above)
Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:265]: Yeah :)

**Review by gunnarmorling — APPROVED** [review]
LGTM now, thanks! Will merge and do some small last polishing.

File: documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc (hunk 10 above)
Comment (rk3rn3r) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:64]: @TovaCohen  already created https://issues.redhat.com/browse/FUSEDOC-4157 to handle adding and styling of that docs section. It's also already annotated for the downstream module handling.

Comment (gunnarmorling) [reply documentation/modules/ROOT/pages/configuration/topic-auto-create-config.adoc:64]: Ah, just seeing below this issue already exists. All good then.

**Comment by gunnarmorling** [comment]
> @TovaCohen <https://github.com/TovaCohen> already created
https://issues.redhat.com/browse/FUSEDOC-4157 to handle adding and styling
of that docs section. It's also already annotated for the downstream module
handling.

Yepp, excellent. Thx!

>

**Pull request merged**